lib/rules/no-constant-binary-expression.js

```diff
@@ -0,0 +1,442 @@
+/**
+ * @fileoverview Rule to flag constant comparisons and logical expressions that always/never short circuit
+ * @author Jordan Eldredge <https://jordaneldredge.com>
+ */
+
+"use strict";
+
+const { isNullOrUndefined } = require("./utils/ast-utils");
+
+// Object keys which, if redefined, could impact how equality is computed.
+const OBJECT_EQUALITY_KEYS = new Set(["valueOf", "toString", "__proto__"]);
+const NUMERIC_OR_STRING_BINARY_OPERATORS = new Set(["+", "-", "*", "/", "%", "|", "^", "&"]);
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Test if an AST node has a statically knowable constant truthiness. Meaning,
+ * it will always coerce to either `true` or `false`.
+ * @param {ASTNode} node The AST node being tested.
+ * @returns {boolean} Does `node` have constant truthiness?
+ */
+function hasConstantTruthiness(node) {
```

Comment:
Thoughts on reusing [`function isConstant`](https://github.com/eslint/eslint/blob/d041f345cdf0306f15faa2f305fe1d21ef137eb1/lib/rules/no-constant-condition.js#L123-L131) from `no-constant-condition`?

It looks like both functions are meant to determine the same thing, which is whether an expression has constant truthiness, so perhaps we could have only one function in a shared place (`ast-utils`) and use it from both rules.

That would catch more cases in `no-constant-binary-expression`. For example, the current version of `no-constant-binary-expression` doesn't report any of these:

```js
-1 || foo; 

1 + 2 || f; 

`a${1}b` === "c" || foo;

a || true || b;
```

While the `no-constant-condition` rule reports same expressions as constant conditions:

```js
-1 ? foo : bar; // no-constant-condition error

1 + 2 ? foo : bar; // no-constant-condition error

`a${1}b` === "c" ? foo : bar; // no-constant-condition error

a || true || b ? foo : bar; // no-constant-condition error
```

(we could add the `Boolean` and `undefined` checks from this implementation to the shared function)

Reply:
We shouldn’t have one rule reach into another rule to use a function. If these two functions do indeed intend to determine the same thing (@captbaritone can comment on that), then moving it into ast-utils makes sense.

Reply:
I think this is a great point. `isConstant` goes to greater effort to detect cases where you can determine that truthiness is constant based based on having constant subexpressions and is therefore it is able to catch more issues.

I agree that it would make sense to extract that into a helper function and use it in both places. But this also makes me think that `hasConstantNullishness` should probably be equally aggressive in trying to understand the impact of subexpressions. I'll need to take a little more time to see how/if there's any way for those two cases to share some of this logic.

Reply:
I'm still trying to make sense of `isConstant` and what exactly we are checking when `inBooleanPosition` is `false`. I finally realized that at least part of my confusion was due to the fact that there are some edge cases which it does not handle correctly. I've reported those as https://github.com/eslint/eslint/issues/15467.

I think that currently `isConstant` with `inBooleanPosition` set to `false` is _supposed_ to check if a value has constant truthiness, _even when that value is being first coerced to some other type_ (string/number/...). I'm not sure that's a function which is feasible to write correctly. Instead, I suspect we will want individual functions that handle the `string` and `number` cases, or something along those lines.

Reply:
With https://github.com/eslint/eslint/pull/15486 merged, I think it _does_ make sense to extract `isConstant` (and therefore its helper functions) into a utility shareable by multiple rules. Should I do that as part of this PR, or a separate PR?

Reply:
@nzakas Thoughts on how/if to go about moving `isConstant` from the `no-constant-condition` rule (and a few functions it depends upon) into a util?

Reply:
It's fine to extract `isConstant` (and its helper functions) to `rules/ast-utils.js` as a part of this PR.

Reply:
> It's fine to extract `isConstant` (and its helper functions) to `rules/ast-utils.js` as a part of this PR.

Also, since there are plenty of tests for `no-constant-condition` rule, which basically just calls `isConstant`, I think separate unit tests for the extracted `astUtils.isConstant` are not necessary at the moment, so we can just copy & paste the code.